fuzz/fuzz_targets/cranelift-fuzzgen.rs

```diff
@@ -19,6 +24,90 @@ use smallvec::smallvec;
 
 const INTERPRETER_FUEL: u64 = 4096;
 
+/// Gather statistics about the fuzzer executions
+struct Statistics {
+    /// All inputs that we tried
+    pub total_inputs: AtomicU64,
+    /// Inputs that fuzzgen can build a function with
+    /// This is also how many compiles we executed
+    pub valid_inputs: AtomicU64,
+
+    /// Total amount of runs that we tried in the interpreter
+    /// One fuzzer input can have many runs
+    pub total_runs: AtomicU64,
+    /// How many runs were successful?
+    /// This is also how many runs were run in the backend
+    pub run_result_success: AtomicU64,
+    /// How many runs resulted in a timeout?
+    pub run_result_timeout: AtomicU64,
+    /// How many runs ended with a trap?
+    pub run_result_trap: HashMap<CraneliftTrap, AtomicU64>,
+}
+
+impl Statistics {
+    pub fn print(&self) {
+        let total_inputs = self.total_inputs.load(Ordering::SeqCst);
+        if total_inputs != 50000 {
+            return;
+        }
+
+        let valid_inputs = self.valid_inputs.load(Ordering::SeqCst);
+        let total_runs = self.total_runs.load(Ordering::SeqCst);
+        let run_result_success = self.run_result_success.load(Ordering::SeqCst);
+        let run_result_timeout = self.run_result_timeout.load(Ordering::SeqCst);
+
+        println!("== FuzzGen Statistics  ====================");
+        println!("Total Inputs: {}", total_inputs);
+        println!(
+            "Valid Inputs: {} ({:.1}%)",
+            valid_inputs,
+            (valid_inputs as f64 / total_inputs as f64) * 100.0
+        );
+        println!("Total Runs: {}", total_runs);
+        println!(
+            "Successful Runs: {} ({:.1}% of Total Runs)",
+            run_result_success,
+            (run_result_success as f64 / total_runs as f64) * 100.0
+        );
+        println!(
+            "Timed out Runs: {} ({:.1}% of Total Runs)",
+            run_result_timeout,
+            (run_result_timeout as f64 / total_runs as f64) * 100.0
+        );
```

Comment:
Comparing Traps and Timeouts against runs is a correct but misleading metric.

We have many runs per input.

Every run has a chance to either trap or timeout. However when they do so we drop the entire input instead of just that run. This causes these values to be under represented in terms of percentage.

A better metric would be timeouts / valid inputs. But it also seems like a very wrong metric in its own way. However it would better represent inefficiencies in the fuzzer.

Reply:
I think this is no longer very relevant timeouts are now solidly in 0.0% territory and with #4895 it now makes sense to compare traps with runs since they no longer drop an input.